File: leetcode/Hashtable.cpp
//
// Created by wzq12 on 2022-08-30.
//
#include <vector>
#include <string>
#include <unordered_set>

using namespace std;
/**
 * 41. 缺失的第一个正数
 * @param nums
 * @return
 */
int firstMissingPositive(vector<int>& nums) {

}


/**
 * 242.有效的字母异位词
 * @param s
 * @param t
 * @return
 *
 */
bool isAnagram(string s, string t) {
    if(s.size() != t.size()) return false;
    int record[26] = {0};
    for (int i = 0; i < s.size(); ++i) {
        record[s[i]-'a']++;
    }
    for (int i = 0; i < t.size(); ++i) {
        record[t[i]-'a']--;
    }
    for (int re: record) {
        if (re!=0) return false;
    }
    return true;
}

/**
 * 349. 两个数组的交集
 * @param nums1
 * @param nums2
 * @return
 */
vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
    unordered_set<int> result; // 存放结果，之所以用set是为了给结果集去重
    unordered_set<int> nums_set(nums1.begin(),nums1.end());
    for (int num: nums2) {
        if (nums_set.find(num) != nums_set.end()) result.insert(num);
    }
    return vector<int>(result.begin(),result.end());
}

File: leetcode/Hashtable.py
from typing import List


def twoSum(self, nums: List[int], target: int) -> List[int]:
    '''
    1. 两数之和
    :param self:
    :param nums:
    :param target:
    :return:
    '''
    records = dict()
    for i, val in enumerate(nums):
        if target - val in records:
            return [records[target - val], i]
        records[val] = i
    return []


def isAnagram(self, s: str, t: str) -> bool:
    """
    242.有效的字母异位词
    :param self:
    :param s:
    :param t:
    :return:
    """
    record = [0] * 26
    for i in s:
        record[ord(i) - ord("a")] += 1
    for i in t:
        record[ord(i) - ord("a")] -= 1
    for i in range(26):
        if record[i] != 0:
            return False
    return True


def threeSum(self, nums: List[int]) -> List[List[int]]:
    nums.sort()
    result = []
    for i in range(len(nums)):
        if nums[i]>0:
            return result
    if i>0 and nums[i]:
        return

